adaptors/modbus/pkg/physical/converter.go

```diff
@@ -1,36 +1,344 @@
 package physical
 
 import (
+	"encoding/hex"
+	"math"
 	"strconv"
+	"strings"
 
 	"github.com/pkg/errors"
 
 	"github.com/rancher/octopus/adaptors/modbus/api/v1alpha1"
+	"github.com/rancher/octopus/pkg/util/converter"
 )
 
-// doArithmeticOperations helps to calculate the raw value with operations,
-// and returns the calculated raw result in 6 digit precision.
-func doArithmeticOperations(raw float64, operations []v1alpha1.ModbusDeviceArithmeticOperation) (string, error) {
-	if len(operations) == 0 {
+// binaryDecodeString converts the binary string value to a byte array.
+func binaryDecodeString(str string) ([]byte, error) {
+	if len(str)%8 != 0 {
+		return nil, errors.New("the length of binary string value must be a multiple of eight")
+	}
+	var val []byte
+	for si := 0; si < len(str); si = si + 8 {
+		var v byte
+		for sj := si + 7; sj >= si; sj-- {
+			if str[sj] == 0x0031 { // == 1
+				v = v | (0b1 << (7 + si - sj))
+			}
+		}
+		val = append(val, v)
+	}
+	return val, nil
+}
+
+// binaryEncodeToString parses the binary string value from a byte array.
+func binaryEncodeToString(data []byte) string {
+	var valSb = &strings.Builder{}
+	for _, b := range data {
+		for bi := 7; bi >= 0; bi-- {
+			if (b>>bi)&0b1 == 0b1 {
+				valSb.WriteString("1")
+			} else {
+				valSb.WriteString("0")
+			}
+		}
+	}
+	return valSb.String()
+}
```

Comment:
those operations are lack of documentation of how the converting is operated and will make the people hard to read and maintain, can we add some explanatory comments in here, additionally instead of using bit operation are there any convertible method that is more human-readable  e.g using https://stackoverflow.com/questions/48646580/convert-a-bitstring-into-a-byte-array

and another question is why those data converting is necessary to be done in the adaptor instead of the user side, since this will introduce more complexity and the adaptor is more likely to be a data pipeline from my point of view.

Reply:
1. I am not sure which answer is you want to specify, I guess you are pointing to the high-score one. Logically, the current coding is doing the same thing as that answer, but the current one has smaller memory allocation and faster execution efficiency like the last answer of that post. Converting a binary string to a byte array is a very basic computer basis, I will add more comments to describe this method.

2. We still need to display the property value when `kubectl get`: some properties may be non-numerical type, not string/Boolean type, just some variable-length byte array, I think it can be displayed with hex/binary/base64. Generally, the type of property is determined, the user can use `int*/uint*/float*/string/boolean` to indicate it, but if the property is not determined or some complex type, what we can do at present is to keep the corresponding byte array and wait for the user to deal with it.